File: mycodes/102-BinaryTreeLevelOrderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) return res;
        queue.offer(root);
        int rest = queue.size();
        List<Integer> curLevel = new ArrayList<>();
        while (!queue.isEmpty()) {
            if (rest == 0) {
                rest = queue.size();
                res.add(curLevel);
                curLevel = new ArrayList<>();
            }
            TreeNode node = queue.poll();
            curLevel.add(node.val);
            rest--;
            if (node.left != null) queue.offer(node.left);
            if (node.right != null) queue.offer(node.right);
        }
        res.add(curLevel);
        return res;
    }
}

File: mycodes/102-BinaryTreeLevelOrderTraversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrder(root *TreeNode) [][]int {
  if root == nil {
    return [][]int{}
  }
  res := make([][]int, 0)
  q := make([]*TreeNode, 0)
  q = append(q, root)
  size := 1
  cur := make([]int, 0, len(q))
  for len(q) > 0 {
    node := q[0]
    q = q[1:]
    size--
    cur = append(cur, node.Val)
    if node.Left != nil {
      q = append(q, node.Left)
    }
    if node.Right != nil {
      q = append(q, node.Right)
    }
    if size == 0 {
      size = len(q)
      res = append(res, cur)
      cur = make([]int, 0, len(q))
    }
  }
  return res
}
